numba/core/bytecode.py

```diff
@@ -321,45 +361,199 @@ def get_used_globals(self):
+        """
+
+        stack = []
+
+        def pop_and_club_exceptions(entries: list, end_entry):
+            start_entry = stack.pop()
+            start_offset = start_entry.start
+            target_offset = start_entry.target
+            end_offset = end_entry.end
+
+            to_be_removed_entries = [
+                e for e in entries if (e.start >= start_offset and
+                                       e.end <= end_offset and
+                                       e.target == target_offset)
+            ]
+
+            lower_entry_idx = entries.index(to_be_removed_entries[0]) - 1
+            upper_entry_idx = entries.index(to_be_removed_entries[-1]) + 1
+
+            if lower_entry_idx >= 0 and upper_entry_idx < len(entries):
+                lower_entry = entries[lower_entry_idx]
+                upper_entry = entries[upper_entry_idx]
+                if lower_entry.target == upper_entry.target:
+                    entries[lower_entry_idx] = _ExceptionTableEntry(
+                        lower_entry.start,
+                        upper_entry.end,
+                        lower_entry.target,
+                        lower_entry.depth,
+                        upper_entry.lasti)
+                    entries.remove(upper_entry)
+
+            dead_entries = [
+                e for e in entries if e.start == target_offset
+            ]
+
+            entries = [e for e in entries if e not in to_be_removed_entries]
+            entries = [e for e in entries if e not in dead_entries]
+            return entries
+
+        for offset, inst in self.table.items():
+            # Check start of pattern, three instructions.
+            # Work out the index of the insteruction.
+            index = self.ordered_offsets.index(offset)
+            # If there is a BUILD_{LIST, MAP, SET} instruction at this
+            # location.
+            curr_inst = self.table[self.ordered_offsets[index]]
+            if curr_inst.opname in ("BUILD_LIST", "BUILD_MAP", "BUILD_SET"):
+                # Check if there are actually two more instructions beyond
+                # this instruction in the table.
+                if len(self.ordered_offsets) <= index + 2:
+                    continue
+                # Check if the BUILD_{LIST, MAP, SET} instruction is followed
+                # by a SWAP(2).
+                next_inst = self.table[self.ordered_offsets[index + 1]]
+                if not next_inst.opname == "SWAP" and next_inst.arg == 2:
+                    continue
+                next_inst = self.table[self.ordered_offsets[index + 2]]
+                # Check if the SWAP is followed by a FOR_ITER
+                if not next_inst.opname == "FOR_ITER":
+                    continue
+
+                # This pattern should be a start of an exception block
+                entry = [e for e in entries if e.start == offset]
+                if not entry:
+                    continue
```

Comment:
performance nitpick

If that's the case, we should just loop over all the exception-entries instead of every instructions in the loop. There must be a lot less exception-entries than instructions.

Reply:
This was fixed in 0836b63048646a6d3d4ac91276468ea989d201a4